### ISSUE TITLE
double Junit5 test execution when selecting test

### ISSUE THREAD

**Issue by rotty3000** [bndtools/bnd#3578]
When selecting tests using junit-platform tester, tests execute twice
![Screenshot from 2019-11-15 23-33-46](https://user-images.githubusercontent.com/146764/68988113-6a9e4180-0800-11ea-906b-20f470146741.png)
Note this might be limited to the bndtools.m2e scenario. I haven't tried with bndtools workspace.
Also, before this can even be tried, you need to have https://github.com/bndtools/bnd/pull/3577

**Comment by rotty3000** [comment]
I'm suspicious was to why there are so many loops through all the bundles here https://github.com/bndtools/bnd/blob/master/biz.aQute.tester.junit-platform/src/aQute/tester/bundle/engine/discovery/BundleSelectorResolver.java#L178-L269

**Comment by rotty3000** [comment]
Why is there a description added for both a bundle and it's fragment? Fragments should essentially be ignored by this parsing. The only action on them might be to find and process their host instead of themselves.

**Comment by rotty3000** [comment]
The issue is that when using class and/or method selectors both the host and fragment bundle produce a sub-request for the same selectors.

**Comment by rotty3000** [comment]
I believe we should either never add a child descriptor in this scenario OR revoke the host's request.

**Comment by bjhargrave** [comment]
Fragments can contain Test-Cases headers that their hosts do not. So fragments must be processed if they are resolved (== attached to a host). It seems wrong that a host's Test-Cases header and its fragment's Test-Cases header would contain the same test classes.

**Comment by bjhargrave** [comment]
> Fragments can contain Test-Cases headers that their hosts do not.

For example see you recent fragment test case support in bnd-maven-plugin.

**Comment by bjhargrave** [comment]
> I believe we should either never add a child descriptor in this scenario OR revoke the host's request.

Both a host and its fragments can contain Test-Cases headers which must be processed. These headers should name tests classes in each bundle and should not overlap.  It seems wrong that a host's Test-Cases header and its fragment's Test-Cases header(s) would contain the same test classes.

**Comment by bjhargrave** [comment]
The processor for Junit 3 and 4 supports Test-Cases headers in both hosts and fragments already.

**Comment by rotty3000** [comment]
What I'm seeing is that the test cases header, only in the fragment, results in a test descriptor for the host and fragment. And this causes the same tests to be run for both descriptors.

```
# Execution Started: [15] org.osgi:org.osgi.test.junit5;1.0.0.201911171703 - [engine:bnd-bundle-engine]/[bundle:org.osgi.test.junit5;1.0.0.201911171703]
# Execution Started: [14] org.osgi:org.osgi.test.junit5;1.0.0.201911171703 - [engine:bnd-bundle-engine]/[bundle:org.osgi.test.junit5;1.0.0.201911171703]/[fragment:org.osgi.test.junit5-test;1.0.0.201911171703]
```

**Comment by rotty3000** [comment]
> And this causes the same tests to be run for both descriptors.
See the image of Eclipse I pasted above.

**Comment by rotty3000** [comment]
Also, when running from the maven cli, and using the `-Dtest=` argument, no tests are run.

**Comment by rotty3000** [comment]
here's an example from cli with trace:
`mvn verify -Pbnd-next -Dtest=org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1`
```
# running in main thread
# test cases org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1
# run unresolved true
# using /home/rotty/projects/osgi-test/org.osgi.test.junit5/target/test-reports/test, path: /home/rotty/projects/osgi-test/org.osgi.test.junit5/target/test-reports/test
# received names of classes to test org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1
# testing request {configurationParameters: LauncherConfigurationParameters [aQute.bnd.junit.bundle.engine.checkUnresolved = 'true'], engineFilters: [], postDiscoverFilters: [], selectors: [MethodSelector [className = 'org.osgi.test.junit5.context.BundleContextExtensionExampleTest', methodName = 'test1', methodParameterTypes = '']]}
# TestPlan Execution Started: org.junit.platform.launcher.TestPlan@424ebba3
# Execution Started: Bnd JUnit Platform Bundle Engine - [engine:bnd-bundle-engine]
# Execution Finished: Bnd JUnit Platform Bundle Engine - [engine:bnd-bundle-engine] - TestExecutionResult [status = SUCCESSFUL, throwable = null]
# TestPlan Execution Finished: org.junit.platform.launcher.TestPlan@424ebba3

Test run finished after 5 ms
[         1 containers found      ]
[         0 containers skipped    ]
[         1 containers started    ]
[         0 containers aborted    ]
[         1 containers successful ]
[         0 containers failed     ]
[         0 tests found           ]
[         0 tests skipped         ]
[         0 tests started         ]
[         0 tests aborted         ]
[         0 tests successful      ]
[         0 tests failed          ]

[INFO] No Errors
...
```

**Comment by rotty3000** [comment]
It seems that when I use the `-Dtest=` engines are not found and so the child descriptors get pruned off by the junit5 launcher.

**Comment by bjhargrave** [comment]
> What I'm seeing is that the test cases header, only in the fragment, results in a test descriptor for the host and fragment.

I am suspecting that getHeaders on the host is returning the Test-Cases from the fragment. So we are seeing the same Test-Cases header from the host and the fragment. We will have this same issue in the Junit 3/4 support as well. We may need to change from using getHeaders to read the Test-Cases header to creating a Manifest object from the META-INF/MANIFEST.MF entry in the host or fragment to make sure we always observe only what it actually in the bundle.

I made #3579 to fix this.

**Comment by bjhargrave** [comment]
> It seems that when I use the -Dtest= engines are not found and so the child descriptors get pruned off by the junit5 launcher.

Perhaps this should be a separate issue?

**Comment by rotty3000** [comment]
@bjhargrave with this fix, I now get the same behaviour in the cli as I do in Eclipse when specifying a single test.

**Comment by kriegfrj** [comment]
> @bjhargrave with this fix, I now get the same behaviour in the cli as I do in Eclipse when specifying a single test _method_.

What you're now seeing is an instance of #3581.

**Comment by kriegfrj** [comment]
On a slightly unrelated issue, from the screenshot, it looks like both your host ([15]) and fragment ([14]) have the same name, which is probably confusing the issue a bit. I might raise a separate issue for this.

**Comment by kriegfrj** [comment]
> I'm suspicious was to why there are so many loops through all the bundles here https://github.com/bndtools/bnd/blob/master/biz.aQute.tester.junit-platform/src/aQute/tester/bundle/engine/discovery/BundleSelectorResolver.java#L178-L269

Yes, it looks a bit untidy. There are many different possible combinations that need to be taken into account (unresolved test bundles, fragments, unattached fragments, test bundles, non-test bundles, unresolved non-test bundles, etc).

I tried doing it all in one pass, but this doesn't change the number of different combinations - it just means that your "leaf" execution in the iteration contains all of the conditional tests instead of in the stream filter.

In the end, I settled for something that works, knowing that it might be possible to do something neater, but also knowing we would also have a set of test cases with good coverage which lay the foundation for attempting to make it neater. It is possible we could use `BundleTracker`s instead of the iterators, for example.
